text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,639 @@
+producer or consumer instrumentation, and it should not cause any changes to
+the relationships of producer and consumer spans. Furthermore, it should be
+possible to provide tracing instrumentation as an out-of-the-box experience
+from messaging SDKs, without requiring any additional custom instrumentation
+from the user.
+
+This OTEP aims at proposing consistent guidelines for creating spans that model
+the stages of the messaging flow, and correlating those in a way so that the
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Terminology
+
+For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).
+
+## Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..-> INTERMEDIARY
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget scenarios, or when
+   settlement happens on the broker), the settlement stage does not exist.
+
+The semantic conventions described below define how to model those stages with
+spans.
+
+## Trace structure
+
+### Producer
+
+Producers are responsible for injecting a creation context into a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus instrumentation libraries and
+auto-instrumentation should default to creating a unique and distinct context
+per message, but may support configuration or other ways to change this default
+behavior.
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
+related "Publish" span) can either refer to a single message or a batch of
+multiple messages.
+
+It is recommended to create a "Create" span for every single message. "Create"
+spans can be created during the "Publish" operation as children of the
```

Comment:
this is confusing. If Publish is a reflection of the batching (under which trace id?) and Create spans are children then they will be all under the same trace id.

Reply:
> "Create" spans can be created during the "Publish" operation as children of the "Publish" span.

In this case, "Create" and "Publish" spans will have the same trace id (that's a requirement for a parent/child relationship).

> Alternatively, "Create" spans can be created independently of the "Publish" operation. In this case, the "Publish" span should link to the "Create" spans.

In this case, "Create" and "Publish" spans can either have the same or different trace ids. In practice, this will likely mean that a user will manually attach a create context to a message _before_ it is passed to the "Publish" operation.

Both scenarios are valid, and we shouldn't prevent or discourage any of them.